Clair3

robert-a-forsyth opened the pull request:
## PR checklist

Closes #XXX <!-- If this PR fixes an issue, please link it here! -->
Adds Clair3 module for germline small variant calling

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [x] `nf-core modules test <MODULE> --profile docker`
    - [x] `nf-core modules test <MODULE> --profile singularity`
    - [x] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

robert-a-forsyth commented:
I need to upload test data to the nf-core test data repository, it's unclear to me if I should add it in this pull request. The website says to contact the nf-core modules slack channel, which I will do now. I have removed all TODO statements, except the one telling myself to change the path of the test data to the location of the newly uploaded test data

SPPearce commented:
> I need to upload test data to the nf-core test data repository, it's unclear to me if I should add it in this pull request. The website says to contact the nf-core modules slack channel, which I will do now. I have removed all TODO statements, except the one telling myself to change the path of the test data to the location of the newly uploaded test data

The data lives in the modules branch of the test-datasets repo, so would definitely need to be a separate PR.

robert-a-forsyth commented:
> > I need to upload test data to the nf-core test data repository, it's unclear to me if I should add it in this pull request. The website says to contact the nf-core modules slack channel, which I will do now. I have removed all TODO statements, except the one telling myself to change the path of the test data to the location of the newly uploaded test data
> 
> The data lives in the modules branch of the test-datasets repo, so would definitely need to be a separate PR.

Opened a pull request just now at (https://github.com/nf-core/test-datasets/pull/1518). I'll remove the comments when that one is approved and change the test set to its location

famosab commented on modules/nf-core/clair3/main.nf:
```diff
@@ -0,0 +1,54 @@
+    def prefix = task.ext.prefix ?: "${meta.id}"
+
+    """
+    touch ${prefix}.phased_merge_output.vcf.gz
```
```suggestion
    echo "" | gzip > ${prefix}.phased_merge_output.vcf.gz
```

famosab commented on modules/nf-core/clair3/main.nf:
```diff
@@ -0,0 +1,54 @@
+    tuple val(meta3), path(fai)
+
+    output:
+    tuple val(meta), path("*merge_output.vcf.gz"), emit: vcf
+    path "versions.yml"           , emit: versions
```
can you also output the vcf index? (tbi)
and align the emit statements :)

famosab commented on modules/nf-core/clair3/main.nf:
```diff
@@ -0,0 +1,54 @@
+
```
```suggestion
```

famosab commented on modules/nf-core/clair3/main.nf:
```diff
@@ -0,0 +1,54 @@
+    stub:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+
```
```suggestion
```

famosab commented on modules/nf-core/clair3/meta.yml:
```diff
@@ -0,0 +1,57 @@
+        description: |
+          Groovy Map containing sample information
+          e.g. `[ id:'sample1', single_end:false ]`
+  
```
```suggestion
```

famosab commented on modules/nf-core/clair3/meta.yml:
```diff
@@ -0,0 +1,57 @@
+          - edam: "http://edamontology.org/format_25722"
+          - edam: "http://edamontology.org/format_2573"
+          - edam: "http://edamontology.org/format_3462"
+          
```
```suggestion
```

famosab commented on modules/nf-core/clair3/tests/main.nf.test:
```diff
@@ -0,0 +1,83 @@
+        then {
+            assertAll(
+                { assert process.success },
+                { assert snapshot(process.out).match() }
```
You will probably need to use the nft-vcf pulgin and do something like

path(process.out.vcf.get(0).get(1)).vcf.variantsMD5

famosab commented on modules/nf-core/clair3/tests/main.nf.test.snap:
```diff
@@ -0,0 +1,72 @@
+                            "id": "test",
+                            "single_end": false
+                        },
+                        "/vsc-hard-mounts/leuven-data/351/vsc35115/modules/.nf-test/tests/d7ade6ea9a7c1a5c1b6266e454543703/work/1b/0e329e076b12f490aa0488da4aef6c/test.phased_merge_output.vcf.gz"
```
This should not be such a path but insteat just the vcf file

famosab replied:
the more you know

famosab commented on modules/nf-core/clair3/main.nf:
```diff
@@ -0,0 +1,54 @@
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    
```
```suggestion
```

famosab commented on modules/nf-core/clair3/main.nf:
```diff
@@ -0,0 +1,54 @@
+        --output=\$PWD \\
+        --platform=$platform \\
+        --model=$model \\
+        $args
```
```suggestion
        $args

```

famosab commented on modules/nf-core/clair3/main.nf:
```diff
@@ -0,0 +1,54 @@
+        --bam_fn=$bam \\
+        --ref_fn=$fasta \\
+        --threads=$task.cpus \\
+        --output=\$PWD \\
```
```suggestion
        --output=${prefix}.phased_merge_output.vcf.gz \\
```

robert-a-forsyth replied:
The output flag is unintuitively the output directory path prefix

famosab replied:
```suggestion
        --output=. \\
```

famosab commented on modules/nf-core/clair3/main.nf:
```diff
@@ -0,0 +1,54 @@
+    tuple val(meta3), path(fai)
+
+    output:
+    tuple val(meta), path("*merge_output.vcf.gz"), emit: vcf
```
```suggestion
    tuple val(meta), path("*.phased_merge_output.vcf.gz"), emit: vcf
```

robert-a-forsyth commented:
Bit of an issue here, Clair3 requires a model path (a folder containing all the model files). I've uploaded this on the github test data, but I now realize that pulling from there doesn't work for folders. So I need some way of downloading the individual files from github in nf-test and passing their parent folder as input to the process.

If you have an ideas to solve this, they'd be deeply appreciated. I've had limited success with the nf-test documentation

famosab commented:
> Bit of an issue here, Clair3 requires a model path (a folder containing all the model files). I've uploaded this on the github test data, but I now realize that pulling from there doesn't work for folders. So I need some way of downloading the individual files from github in nf-test and passing their parent folder as input to the process.

How about making a tar.gz folder upload that to the test-datasets repo and then use untar to stage it for the tests. Like done here: https://github.com/nf-core/modules/blob/6b8947dd401a58b08ca19f1952dc1182d66e4d89/subworkflows/nf-core/fastq_align_dedup_bwameth/tests/gpu.nf.test#L22

robert-a-forsyth commented:
> > Bit of an issue here, Clair3 requires a model path (a folder containing all the model files). I've uploaded this on the github test data, but I now realize that pulling from there doesn't work for folders. So I need some way of downloading the individual files from github in nf-test and passing their parent folder as input to the process.
> 
> How about making a tar.gz folder upload that to the test-datasets repo and then use untar to stage it for the tests. Like done here:
> 
> https://github.com/nf-core/modules/blob/6b8947dd401a58b08ca19f1952dc1182d66e4d89/subworkflows/nf-core/fastq_align_dedup_bwameth/tests/gpu.nf.test#L22

Ah that's brilliant, I hadn't realized you could do that in nf-test. Thank you so much!

robert-a-forsyth commented:
Tried to fix the suggestions, added the data. I've changed phase_merge_output.vcf to optional, since you need to specific with arguments that you will be phasing. But it is now emitted

SPPearce commented on modules/nf-core/clair3/main.nf:
```diff
@@ -0,0 +1,57 @@
+        'biocontainers/clair3:1.0.10--py39hd649744_1' }"
+
+    input:
+    tuple val(meta), path(bam), path(bai), path(model),val(platform)
```
```suggestion
    tuple val(meta), path(bam), path(bai), path(model), val(platform)
```

SPPearce commented on modules/nf-core/clair3/main.nf:
```diff
@@ -0,0 +1,57 @@
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    echo "" | gzip > ${prefix}.phased_merge_output.vcf.gz
+    echo "" | gzip > ${prefix}.phased_merge_output.vcf.gz.tbi
```
Wouldn't actually be a problem, but it is only the .gz files that need to be gzipped.
```suggestion
    touch ${prefix}.phased_merge_output.vcf.gz.tbi
```

SPPearce commented on modules/nf-core/clair3/main.nf:
```diff
@@ -0,0 +1,57 @@
+    echo "" | gzip > ${prefix}.phased_merge_output.vcf.gz
+    echo "" | gzip > ${prefix}.phased_merge_output.vcf.gz.tbi
+    echo "" | gzip > ${prefix}.merge_output.vcf.gz
+    echo "" | gzip > ${prefix}.merge_output.vcf.gz.tbi
```
```suggestion
    touch ${prefix}.merge_output.vcf.gz.tbi
```

SPPearce commented on modules/nf-core/clair3/meta.yml:
```diff
@@ -0,0 +1,118 @@
+name: "clair3"
+description: germline small variant caller
+keywords:
+- sort
+- example
+- genomics
```
Better keywords please :)

SPPearce commented on modules/nf-core/clair3/meta.yml:
```diff
@@ -0,0 +1,118 @@
+# yaml-language-server: $schema=https://raw.githubusercontent.com/nf-core/modules/master/modules/meta-schema.json
+name: "clair3"
+description: germline small variant caller
```
```suggestion
description: Clair3 is a germline small variant caller for long-reads
```

SPPearce commented on modules/nf-core/clair3/meta.yml:
```diff
@@ -0,0 +1,118 @@
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. `[ id:'sample1', single_end:false ]`
```
```suggestion
        e.g. `[ id:'sample1' ]`
```
Can we remove all the single_end:false if possible, as they aren't relevant here.

SPPearce reviewed (changes requested):
Almost ready to go I think.

SPPearce commented on modules/nf-core/clair3/main.nf:
```diff
@@ -0,0 +1,57 @@
+    touch ${prefix}.phased_merge_output.vcf.gz.tbi
+    echo "" | gzip > ${prefix}.merge_output.vcf.gz
+    touch ${prefix}.merge_output.vcf.gz.tbi
+    cat <<-END_VERSIONS > versions.yml
```
```suggestion

    cat <<-END_VERSIONS > versions.yml
```

SPPearce reviewed (approved):
Ok, I'm happy.
